Adafruit_MQTT.cpp

```diff
@@ -562,15 +562,33 @@ Adafruit_MQTT_Subscribe *Adafruit_MQTT::handleSubscriptionPacket(uint16_t len) {
     return NULL;
   }
 
-  // Parse out length of packet.
-  // NOTE: This includes data in the variable header and the payload.
-  uint16_t remainingLen = len - 4; // subtract the 4 header bytes
-  uint16_t const topicoffset = packetAdditionalLen(remainingLen);
-  uint16_t const topicstart = topicoffset + 4;
-
-  topiclen = int((buffer[2 + topicoffset]) << 8 | buffer[3 + topicoffset]);
-  DEBUG_PRINT(F("Looking for subscription len "));
-  DEBUG_PRINTLN(topiclen);
+  // Parse MQTT Remaining Length field (variable length encoding).
+  // This field can be 1-4 bytes, so we must parse it byte-by-byte.
+  // Each byte uses the lower 7 bits for data, and the MSB as a continuation
+  // flag. This loop advances 'offset' to the first byte after the remaining
+  // length field.
+  uint32_t multiplier = 1;
+  uint32_t value = 0;
+  uint8_t encodedByte;
+  uint8_t offset = 1; // Start after the fixed header byte
+  do {
+    encodedByte = buffer[offset++];
+    value += (encodedByte & 127) * multiplier;
+    multiplier *= 128;
+    // Infinite loop protection: MQTT spec allows max 4 bytes for this field
+    if (offset > 5) { // 1 (header) + 4 (max length bytes)
+      DEBUG_PRINTLN(
+          F("Bad MQTT packet: Remaining Length field too long / malformed"));
+      return NULL;
+    }
+  } while ((encodedByte & 128) != 0);
+
```

Comment:
Perform a check for safety at this point - the maximum remaining length should be less than `268,435,455`.

Reply:
The safety check is performed a little earlier, where the algorithm exits (returns NULL) if the cycle is performed more than 4 times (line 579).
On the fifth run the function will exit and Value will be unused (it's unused anyway - we only use the offset).
The maximum value is therefore capped at 268435455 algorithmically.
<img width="854" height="198" alt="image" src="https://github.com/user-attachments/assets/f42f58dc-83c0-41db-b828-cceef30eea2b" />

It seems the [spec suggests](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718023) from 0 to 268435455 inclusive, so I think this is good to go.